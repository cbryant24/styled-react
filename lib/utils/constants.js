import * as styledSystem from 'styled-system';
const { compose, system } = styledSystem;

//TODO: do i need these? is it common enough? possibly remove check css properties lists 
const APPEARANCE = system({
  appearance: true
});

export const CURSOR = system({
  cursor: true
});

//END TODO

// fontSizeModules custom theme property see https://styled-system.com/custom-props
// see https://css-tricks.com/rem-global-em-local/ for global and local sizing design
const FONTSIZESMODULES = system({
  fontSizeModule: {
    property: 'fontSize',
    scale: 'fontSizesModule'
  }
});

// TODO: create basic more restrictive html css elements
// group most common styled-system css properties togethers for theming 
export const COMMON = compose(
  styledSystem.space,
  styledSystem.color,
  styledSystem.layout,
  styledSystem.zIndex,
  styledSystem.position,
  APPEARANCE,
  FONTSIZESMODULES,
  CURSOR
);

// TODO: create basic more restrictive html css border elements
// styled-system border properties
export const BORDER = compose(
  styledSystem.border,
  styledSystem.shadow
);


// styled-system complete transition elements with all properties
export const TRANSITION = system({
  transition: true,
  transitionDelay: true,
  transitionDuration: true,
  transitionProperty: true,
  transitionTimingFunction: true
});

// styled-system complete grid elements with all properties
export const GRID = compose(
  styledSystem.grid,
  system({
    gridRowStart: true,
    gridRowEnd: true,
    gridColumnStart: true,
    gridColumnEndt: true,
    itemName: true
  })
);

//TODO: add more restricitive html css typography elements
// styled-system complete typography css properties 
export const TYPOGRAPHY = compose(
  styledSystem.typography,
  system({
    textTransform: true,
    textDecoration: true
  })
)

export const FLEX       = styledSystem.flexbox;
export const SHADOW     = styledSystem.shadow;
export const BACKGROUND = styledSystem.background;

// combine all styled-system and and added styled-system custom props
export const ALL = styledSystem.compose(
  BORDER,
  BACKGROUND,
  COMMON,
  FLEX,
  GRID,
  SHADOW,
  TYPOGRAPHY,
  TRANSITION
);
