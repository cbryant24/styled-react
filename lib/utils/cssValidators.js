import { transitionTypes, cssListArr } from './cssHelpers';

// TODO: ADD MORE VALIDATORS TO THIS FILE 
export const validateAnimation = animation => {
  if (!animation) return;

  const checkForValidDuration = duration => {
    return typeof duration === 'number' && duration >= 0;
  };

  const checkForValidIteration = iteration => {
    const rounds = parseInt(iteration, 10);
    return (
      (typeof rounds === 'number' && rounds >= 1) || iteration === 'infinite'
    );
  };

  const checkForValidCSSProperty = property => {
    return cssListArr.includes(property);
  };

  // Check of an in animation
  if ('in' in animation) {
    // Check if that animation has also a duration
    if (!('duration_in' in animation)) {
      throw new TypeError(
        'If you have an in animation you need to specify a duration for that animation'
      );
    } else {
      // Check for Valid Animation Duration In
      if (!checkForValidDuration(animation.duration_in)) {
        throw new TypeError(
          `${
            animation.duration_in
          } is not a valid duration in for an animation`
        );
      }
    }
    // Check if this in animation is going to iterate
    if ('iteration' in animation) {
      // Check for Valid Iteration
      if (!checkForValidIteration(animation.iteration)) {
        throw new TypeError(
          `${
            animation.iteration
          } is not a valid type of iteration property. Should be real number or the 'inifite' literal`
        );
      }
    }
  } else {
    if ('delay_between' in animation) {
      throw new TypeError(
        `You cannot have a delay between in and out animations if you're missing any of them`
      );
    }
  }
  // Check of an out animation
  if ('out' in animation) {
    // Check if that animation has also a duration
    if (!('duration_out' in animation)) {
      throw new TypeError(
        'If you have an out animation you need to specify a duration for that animation'
      );
    }
  } else {
    if ('delay_between' in animation) {
      throw new TypeError(
        `You cannot have a delay between in and out animations if you're missing any of them`
      );
    }
  }
  // Check for both animations (in and out) at the same time
  if ('in' in animation && 'out' in animation) {
    // Check if there's also a delay between those animations
    if (!('delay_between' in animation)) {
      throw new TypeError(
        'If you have an in animation and an out animation you need to specify a delay between those animations'
      );
    }
    // Check for Valid Animation Duration In
    if (!checkForValidDuration(animation.duration_out)) {
      throw new TypeError(
        `${
          animation.duration_out
        } is not a valid duration out for an animation`
      );
    }
  }
  // Check for a continuous animation
  if ('continuous' in animation) {
    if ('duration_continuous' in animation) {
      // Check for Valid Animation Duration In
      if (!checkForValidDuration(animation.duration_continuous)) {
        throw new TypeError(
          `${
            animation.duration_continuous
          } is not a valid duration for a continuous animation`
        );
      }
    } else {
      throw new TypeError(
        `${
          animation.duration_out
        } is not a valid duration out for an animation`
      );
    }
  }
};

export const validateTransitions = transition => {
  if (!transition) return;

  const checkForValidTransitionType = transitionType => {
    return transitionTypes.includes(transitionType);
  };
  // debugger
  if (!('type' in transition)) {
    throw new TypeError(
      `You're missing the type of transition property. Eg: hover, focus, blur, active, ...`
    );
  }
  // Check for Valid Transition Type
  if (!checkForValidTransitionType(transition.type)) {
    throw new TypeError(
      `${transition.type} is not a valid type of transition`
    );
  }

  const checkForValidFromToObject = fromToObject => {
    return 'property' in fromToObject && 'value' in fromToObject;
  };

  const checkForValidCSSProperty = property => {
    return cssListArr.includes(property);
  };

  // Check for from transition object
  if (!('from' in transition)) {
    throw new TypeError(
      `You're missing the from property of the transition that sets the point to start at`
    );
  }
  // Check if from object is valid and meet the requirements
  if (!checkForValidFromToObject(transition.from)) {
    throw new TypeError(
      `${JSON.stringify(
        transition.from
      )} is not a valid transition FROM object. It needs to have the following structure: { property: string, value: string || Number }`
    );
  }
  // Check if the from css property is valid
  if (!checkForValidCSSProperty(transition.from.property)) {
    throw new TypeError(
      `${transition.from.property} is not a valid CSS property at FROM object`
    );
  }
  // Check for to transition object
  if (!('to' in transition)) {
    throw new TypeError(
      `You're missing the to property of the transition that sets the point of where to end at`
    );
  }
  // Check if to object is valid and meet the requirements
  if (!checkForValidFromToObject(transition.to)) {
    throw new TypeError(
      `${JSON.stringify(
        transition.from
      )} is not a valid transition TO object. It needs to have the following structure: { property: string, value: string || Number }`
    );
  }
  // Check if the to css property is valid
  if (!checkForValidCSSProperty(transition.to.property)) {
    throw new TypeError(
      `${transition.from.property} is not a valid CSS property at TO object`
    );
  }
};