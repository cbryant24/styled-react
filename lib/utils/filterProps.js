import { omit } from 'lodash';

export const filterProps = props => {
  // debugger
  return omit(props, [
    'theme',
    'xs',
    'sm',
    'md',
    'lg',
    'xl',
    'color',
    'bg',
    'fill',
    'fontSize',
    'fontFamily',
    'fontStyle',
    'f',
    'image',
    'maxWidth',
    'bold',
    'regular',
    'caps',
    'wrap',
    'size',
    'height',
    'boxShadowSize',
    'remSizes',
    "templateRows",
    "templateColumns",
    "alignSelf",
    "justifySelf",
    "gridColumn",
    "gridRow",
    "alignContent",
    "gridGap",
    "textTransform",
    "minHeight",
    "flexDirection",
    "borderColor",
    "justifyContent",
    "borderRadius",
    "borderWidth",
    "bgAnimation",
    "backgroundColor",
    "textShadow",
    "zIndex",
    "backgroundImage",
    "backgroundSize",
    "transformOrigin",
    "alignItems",
    "boxShadow",
    "fontSmooth",
    "bgAnimation",
    "windowBorder",
    "bgAnimationGlow",
    "textAlign",
    "flexGrow",
    "foucsBoxShadowColor",
    "focusColor",
    "borderFocus",
    "maxHeight",
    "foucsBoxShadowColor",
    "focusColor",
    "borderFocus",
    "error",
    "minHeight",
    "fontSmooth",
    "clipPath",
    "transformOrigin",
    "transform",
    "content",
    "all",
    "animation",
    "animationDelay",
    "animationDirection",
    "animationDuration",
    "animationFillMode",
    "animationIterationCount",
    "animationName",
    "animationPlayState",
    "animationTimingFunction",
    "appearance",
    "azimuth",
    "backdropFilter",
    "backfaceVisibility",
    "backgroundAttachment",
    "backgroundBlendMode",
    "backgroundClip",
    "backgroundOrigin",
    "backgroundPositionX",
    "backgroundPositionY",
    "blockSize",
    "borderBlockEnd",
    "borderBlockEndColor",
    "borderBlockEndStyle",
    "borderBlockEndWidth",
    "borderBlockStart",
    "borderBlockStartColor",
    "borderBlockStartStyle",
    "borderBlockStartWidth",
    "borderBottomColor",
    "borderBottomLeftRadius",
    "borderBottomRightRadius",
    "borderBottomStyle",
    "borderBottomWidth",
    "borderCollapse",
    "borderImage",
    "borderImageOutset",
    "borderImageRepeat",
    "borderImageSlice",
    "borderImageSource",
    "borderImageWidth",
    "borderInlineEnd",
    "borderInlineEndColor",
    "borderInlineEndStyle",
    "borderInlineEndWidth",
    "borderInlineStart",
    "borderInlineStartColor",
    "borderInlineStartStyle",
    "borderInlineStartWidth",
    "borderLeftColor",
    "borderLeftStyle",
    "borderLeftWidth",
    "borderRightColor",
    "borderRightStyle",
    "borderRightWidth",
    "borderSpacing",
    "borderTopColor",
    "borderTopLeftRadius",
    "borderTopRightRadius",
    "borderTopStyle",
    "borderTopWidth",
    "bottom",
    "boxAlign",
    "boxDecorationBreak",
    "boxDirection",
    "boxFlex",
    "boxFlexGroup",
    "boxLines",
    "boxOrdinalGroup",
    "boxOrient",
    "boxPack",
    "boxSizing",
    "breakAfter",
    "breakBefore",
    "breakInside",
    "captionSide",
    "caretColor",
    "clear",
    "clip",
    "columnCount",
    "columnFill",
    "columnRule",
    "columnRuleColor",
    "columnRuleStyle",
    "columnRuleWidth",
    "columnSpan",
    "columnWidth",
    "columns",
    "contain",
    "counterIncrement",
    "counterReset",
    "cursor",
    "direction",
    "displayInside",
    "displayList",
    "displayOutside",
    "emptyCells",
    "filter",
    "flexFlow",
    "float",
    "font",
    "fontFeatureSettings",
    "fontKerning",
    "fontLanguageOverride",
    "fontVariationSettings",
    "fontSizeAdjust",
    "fontStretch",
    "fontSynthesis",
    "fontVariant",
    "fontVariantAlternates",
    "fontVariantCaps",
    "fontVariantEastAsian",
    "fontVariantLigatures",
    "fontVariantNumeric",
    "fontVariantPosition",
    "gridColumnEnd",
    "gridColumnStart",
    "gridRowEnd",
    "gridRowStart",
    "gridTemplate",
    "hyphens",
    "imageOrientation",
    "imageRendering",
    "imageResolution",
    "imeMode",
    "initialLetter",
    "initialLetterAlign",
    "inlineSize",
    "isolation",
    "lineBreak",
    "lineHeightStep",
    "listStyle",
    "listStyleImage",
    "listStylePosition",
    "listStyleType",
    "marginBlockEnd",
    "marginBlockStart",
    "marginInlineEnd",
    "marginInlineStart",
    "markerOffset",
    "mask",
    "maskClip",
    "maskComposite",
    "maskImage",
    "maskMode",
    "maskOrigin",
    "maskPosition",
    "maskRepeat",
    "maskSize",
    "maskType",
    "maxBlockSize",
    "maxInlineSize",
    "minBlockSize",
    "minInlineSize",
    "mixBlendMode",
    "objectFit",
    "objectPosition",
    "offset",
    "offsetAnchor",
    "offsetBlockEnd",
    "offsetBlockStart",
    "offsetInlineEnd",
    "offsetInlineStart",
    "offsetDistance",
    "offsetPath",
    "offsetPosition",
    "offsetRotate",
    "order",
    "orphans",
    "outline",
    "outlineColor",
    "outlineOffset",
    "outlineStyle",
    "outlineWidth",
    "overflow",
    "overflowClipBox",
    "overflowWrap",
    "overflowX",
    "overflowY",
    "padding",
    "paddingBlockEnd",
    "paddingBlockStart",
    "paddingInlineEnd",
    "paddingInlineStart",
    "pageBreakAfter",
    "pageBreakBefore",
    "pageBreakInside",
    "perspective",
    "perspectiveOrigin",
    "pointerEvents",
    "quotes",
    "resize",
    "rubyAlign",
    "rubyMerge",
    "rubyPosition",
    "scrollBehavior",
    "scrollSnapCoordinate",
    "scrollSnapDestination",
    "scrollSnapPointsX",
    "scrollSnapPointsY",
    "scrollSnapType",
    "scrollSnapTypeX",
    "scrollSnapTypeY",
    "shapeImageThreshold",
    "shapeMargin",
    "shapeOutside",
    "tabSize",
    "tableLayout",
    "textAlignLast",
    "textCombineUpright",
    "textDecoration",
    "textDecorationColor",
    "textDecorationLine",
    "textDecorationSkip",
    "textDecorationStyle",
    "textEmphasis",
    "textEmphasisColor",
    "textEmphasisPosition",
    "textEmphasisStyle",
    "textIndent",
    "textJustify",
    "textOrientation",
    "textOverflow",
    "textRendering",
    "textSizeAdjust",
    "textUnderlinePosition",
    "touchAction",
    "transformBox",
    "transformStyle",
    "transition",
    "transitionDelay",
    "transitionDuration",
    "transitionProperty",
    "transitionTimingFunction",
    "unicodeBidi",
    "userSelect",
    "verticalAlign",
    "visibility",
    "whiteSpace",
    "widows",
    "willChange",
    "wordBreak",
    "wordSpacing",
    "wordWrap",
    "writingMode",
    "font-size",
    "fontSize",
    "font-family",
    "fontFamily",
    "width",
    "height",
    "borderstyle",
    "borderStyle",
    "lineHeight",
    "letterSpacing",
    "margin",
    "display",
    "letter-spacing",
    "m",
    "mx",
    "my",
    "p",
    "px",
    "py",
    "border",
    "opacity",
    "border-radius",
    "borderBottom",
    "borderLeft",
    "borderRight",
    "borderTop",
    "background",
    "position",
    "flex-flow",
    "grid-row-start",
    "grid-column-start",
    "grid-row-end",
    "grid-column-end",
    "itemname",
    "focus",
    "::before",
    "::after",
    "fontSizeModule",
    "borderWidth",
    "cursor",
    "errors",
    "top",
    "left",
    "right",
    "bottom",
    "themeStyle",
    "remove",
    "options",
    "h1FontSize",
    "h2FontSize",
    "h3FontSize",
    "h4FontSize",
    "h5FontSize",
    "h6FontSize",
    "paddingLeft",
    "minWidth",
    "fontWeight",
    "mb",
    "mt",
    "before",
    "after",
    "flexWrap",
    "gridTemplateRows",
    "gridTemplateColumns",
    "backgroundRepeat",
    "backgroundPosition",
    "pseudo",
    "marginRight",
    "marginLeft",
    "marginTop",
    "marginBottom",
    "paddingRight",
    "paddingLeft",
    "paddingTop",
    "paddingBottom"
])}
