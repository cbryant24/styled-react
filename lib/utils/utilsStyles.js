import { system } from 'styled-system';
import { themeGet } from '@styled-system/theme-get';
import styledCSS from '@styled-system/css';

import { cssListObj, pseudoClasses, pseudoElements } from './cssHelpers';


// add css properties not included in styled-system library see styled Box
export const addCssProperties = props => {
  const cssProps = {};
  for (let key in props) {
    if(cssListObj[key]) cssProps[key] = true;
  }
  return system({ ...cssProps });
};

//TODO: Add theming access for pseudo elements
// checks props passed from component if psuedo elements exist except for before and after which are handled specially
export const addPseudo = (props, isThemeStyled=false) => {
  if (!props.pseudo) return;

  const elementProps  = Object.getOwnPropertyNames(props),
  pseudoClassProps    = elementProps.filter( prop => pseudoClasses.includes(prop)),
  pseudoElementProps  = elementProps.filter( prop => pseudoElements.includes(prop)),
  pseudo              = {};

  if (!pseudoClassProps && !pseudoElementProps) return;

  pseudoClassProps.forEach( pseudoType => pseudo[`:${pseudoType}`] = props[pseudoType] );
  pseudoElementProps.forEach( pseudoType => pseudo[`::${pseudoType}`] = props[pseudoType]);

  //TODO: add hyphenated pseudo elements/classes pseudoClasses and pseudoElements
  //add hyphenated pseudo elements and classes if present in props
  if (pseudoClassProps.indexOf('focusWithin') >= 0)
    pseudo[':focus-within'] = pseudo[':focusWithin'];

  if (pseudoElementProps.indexOf('firstLine') >= 0)
    pseudo['::first-line'] = pseudo['::firstLine'];

  if (pseudoElementProps.indexOf('firstLetter') >= 0)
    pseudo['::first-letter'] = pseudo['::firstLetter'];

  // if themeStyled return plain pseudo object to invoked function for theme styling changes
  if (isThemeStyled) return pseudo;

  // @styled-system/css converts object written style to css with theming access
  return styledCSS({ ...pseudo });
}

// TODO: DRY addThemeStyle && addThemeStyleAnimation && addThemeStylePsuedoAnimation possibly 
// returning an object with the three necesary values and destructuring them off
// hooks into theme and adds appropriate styling determined by string or array of strings of appropriate theme objects
export const addThemeStyle = props => {
  if (!props.themeStyle) return;

  if (Array.isArray(props.themeStyle)) {
    let combinedStyle = {};

    props.themeStyle.forEach( style => {
      const themeVals = themeGet(style)(props);
      combinedStyle = { ...combinedStyle, ...themeVals }
    });

    if (combinedStyle.pseudo)
      combinedStyle = {...combinedStyle, ...addPseudo(combinedStyle, true)}
    
    const { animation, ...combinedStylesNoAnimation } = combinedStyle;

    return styledCSS(combinedStylesNoAnimation);
  }

  let themedStyles = themeGet(props.themeStyle)(props);
  try {
    if (themedStyles.pseudo) 
    themedStyles = {...themedStyles, ...addPseudo(themedStyles, true)}
  } catch(err) {
    console.log('error with getting theme style invalid theme name');
    return err;
  }


  // allows for passage of special prop `remove` to remove styling for strict control over specific necessary css styling
  if (props.remove)
    themedStyles = removeStyles(themedStyles, props.remove)

  const { animation, ...themedStylesNoAnimation} = themedStyles;

  return styledCSS(themedStylesNoAnimation);
}

export const addThemeStyleAnimation = props => {
  console.log('WE OUTCHEA', props);
  // if (!props.themeStyle) return;

  if (Array.isArray(props.themeStyle)) {
    let combinedStyle = {};

    props.themeStyle.forEach( style => {
      const themeVals = themeGet(style)(props);
      combinedStyle = { ...combinedStyle, ...themeVals }
    });

    return combinedStyle.animation
  }

  let themedStyles = themeGet(props.themeStyle)(props);
  // try {
  //   if (themedStyles.pseudo) 
  //   themedStyles = {...themedStyles, ...addPseudo(themedStyles, true)}
  // } catch(err) {
  //   console.log('error with getting theme style invalid theme name');
  //   return err;
  // }
  console.log("IM IN THE THEMED ANIMATIONs", themedStyles);
  return themedStyles.animation;
}


//TODO: add another property for string/array and removal of styles to object version if remove is not null
// checks if the style is string or array then returns an object with themeStyle property for later parsing in 
// addThemeStyle function  otherwise if object passes on to removeStyle for style removal
export const styleBuildRemove = (styles, remove) => {
  return (
    typeof styles === 'string' || Array.isArray(styles) ? { themeStyle: styles, remove } : 
                                                  remove ? removeStyles(styles, remove) : { ...styles  }
  )
}


// remove styles from object takes array to remove multiple or string to remove one  
// if no value for remove original styles returned
export const removeStyles = (styles, remove) => {
  let removedStyles = styles;

  if (Array.isArray(remove)) {
    remove.forEach( (style, idx) => {
      const { [style]: removed, ...rest} = removedStyles;
      removedStyles = rest;
    })

    return  removedStyles;
  }
  const { [remove]: removed, ...rest } = styles;
  removedStyles = rest;

  return removedStyles
}
