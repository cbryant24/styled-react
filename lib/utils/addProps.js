import React from 'react';
import { filterProps } from './filterProps';

export const addProps = (component, props) => {
  if (!component || !props) return;

  if (Array.isArray(component)) {
    if(Array.isArray(props)) {
      const newComponentsArr = component.map( (component, idx) => {
        return React.cloneElement( component, { ...props[idx]})
      })

      return newComponentsArr;
    }
    const newComponents = component.map( component => {
      return React.cloneElement( component, { ...props });
    })
    return newComponents;
  }

  const newComponent = React.cloneElement(component, { ...props })

  return newComponent;
}