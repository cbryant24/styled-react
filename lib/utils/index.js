import { filterProps } from './filterProps';
import { cssListArr, cssListObj, pseudoClasses, pseudoElements } from './cssHelpers';
import { validateAnimation } from './cssValidators';
import { 
  addCssProperties, 
  addPseudo, 
  addThemeStyle, 
  styleBuildRemove,
  removeStyles,
  addThemeStyleAnimation } from './functions';
import { addProps } from './addProps';
import { 
  COMMON, BORDER, TYPOGRAPHY, 
  FLEX, GRID, TRANSITION,
  SHADOW, ALL
} from './constants';

export {
  addPseudo,
  addProps,
  addThemeStyle,
  addCssProperties,
  addThemeStyleAnimation,
  removeStyles,
  cssListArr,
  cssListObj,
  filterProps,
  pseudoElements,
  pseudoClasses,
  styleBuildRemove,
  validateAnimation,
  ALL,
  COMMON,
  BORDER,
  FLEX,
  GRID,
  TRANSITION,
  TYPOGRAPHY,
  SHADOW,
};