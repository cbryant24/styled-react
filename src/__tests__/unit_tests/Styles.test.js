import React from 'react'
import { shallow, mount } from 'enzyme';
import {
  Div
} from '@cbryant24/styled-react';
import 'jest-styled-components';
import { before } from 'lodash';
import { updateComponent } from '../utils'


let component;


describe('applied styles', () => {

  it('has a div with margin 50%', () => {
    component = shallow( <Div width="75%"/>);
    
    expect(component).toMatchSnapshot();
    expect(component).toHaveStyleRule('width', '75%');
  });

  it('properly displays camelcase css', () => {
    component = shallow( <Div backgroundColor="red"/>);
    expect(component).toMatchSnapshot();
    expect(component).toHaveStyleRule('background-color', 'red');
  });

  it('properly displays destructed object styles', () => {
    const styles = {clipPath: "circle(40%)", height: '200px'};
    component = shallow(<Div {...styles}/>);
    expect(component).toMatchSnapshot();
    expect(component).toHaveStyleRule('clip-path', 'circle(40%)');
    expect(component).toHaveStyleRule('height', '200px');
  });

  describe('pseudo', () => {
    beforeEach(() => {
      component = shallow(
       <Div
        pseudo
        hover={{ 
          fontSize: '2.4rem', 
          height: '200px', 
          width: '300px',
          color: 'blue'
        }}
        transition="font-size 4s linear, height 2s linear, width 3s linear"
       />
      );
    });

    it('has hover styles', () => {
      expect(component).toMatchSnapshot();
      expect(component).toHaveStyleRule('font-size', '2.4rem', {
        modifier: ':hover'
      });
      expect(component).toHaveStyleRule('height', '200px', {
        modifier: ':hover'
      });
      expect(component).toHaveStyleRule('width', '300px', {
        modifier: ':hover'
      });
      expect(component).toHaveStyleRule('transition', 'font-size 4s linear,height 2s linear,width 3s linear');
    });
  });

  describe('animation', () => {
    beforeEach(async () => {
      component = mount(
        <Div
          animation={{
            in: {
              from: { opacity: 0 },
              to: { opacity: 1 }
            },
            duration_in: 3
          }}
        />
      );

      await updateComponent(component, 10);
    });

    it('has all animation css rules', () => {
      expect(component).toMatchSnapshot();
      expect(component).toHaveStyleRule('animation-name', expect.any(String));
      expect(component).toHaveStyleRule('animation-play-state', 'running');
      expect(component).toHaveStyleRule('animation-duration', '3s');
    })
  })
});