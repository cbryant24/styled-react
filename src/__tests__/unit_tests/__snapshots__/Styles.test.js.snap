// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applied styles animation has all animation css rules 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-animation-name: iHtkXv;
  animation-name: iHtkXv;
  -webkit-animation-duration: 3s;
  animation-duration: 3s;
  -webkit-animation-play-state: running;
  animation-play-state: running;
}

<Styled(Styled(Component))
  animation={
    Object {
      "duration_in": 3,
      "in": Object {
        "from": Object {
          "opacity": 0,
        },
        "to": Object {
          "opacity": 1,
        },
      },
    }
  }
>
  <Component
    animation={
      Object {
        "duration_in": 3,
        "in": Object {
          "from": Object {
            "opacity": 0,
          },
          "to": Object {
            "opacity": 1,
          },
        },
      }
    }
    className="c0"
    isA="div"
  >
    <div
      className="c0"
    />
  </Component>
</Styled(Styled(Component))>
`;

exports[`applied styles has a div with margin 50% 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  width: 75%;
}

<Component
  className="c0"
  isA="div"
  width="75%"
/>
`;

exports[`applied styles properly displays camelcase css 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  background-color: red;
}

<Component
  backgroundColor="red"
  className="c0"
  isA="div"
/>
`;

exports[`applied styles properly displays destructed object styles 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-clip-path: circle(40%);
  clip-path: circle(40%);
  height: 200px;
}

<Component
  className="c0"
  clipPath="circle(40%)"
  height="200px"
  isA="div"
/>
`;

exports[`applied styles pseudo has hover styles 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: font-size 4s linear,height 2s linear,width 3s linear;
  transition: font-size 4s linear,height 2s linear,width 3s linear;
  -webkit-transition: font-size 4s linear,height 2s linear,width 3s linear;
  transition: font-size 4s linear,height 2s linear,width 3s linear;
}

.c0:hover {
  font-size: 2.4rem;
  height: 200px;
  width: 300px;
  color: blue;
}

.c0:hover {
  font-size: 2.4rem;
  height: 200px;
  width: 300px;
  color: blue;
}

<Component
  className="c0"
  hover={
    Object {
      "color": "blue",
      "fontSize": "2.4rem",
      "height": "200px",
      "width": "300px",
    }
  }
  isA="div"
  pseudo={true}
  transition="font-size 4s linear, height 2s linear, width 3s linear"
/>
`;
